### PR TITLE
Bug #2580: Initialize and load hwloc topology only once, during startup

### PR DESCRIPTION
**Pull request by evan-charmworks**

### CODE REVIEW

File: src/conv-core/cpuaffinity.C [hunk 1]
```diff
@@ -37,16 +37,12 @@ CpvDeclare(void *, myProcStatFP);
 
 CmiHwlocTopology CmiHwlocTopologyLocal;
 
+static hwloc_topology_t topology, legacy_topology;
```
Comment (stwhite91) [inline src/conv-core/cpuaffinity.C]: A comment here explaining the difference between these two might be good

File: src/conv-core/cpuaffinity.C [hunk 2]
```diff
@@ -64,19 +60,12 @@ void CmiInitHwlocTopology(void)
...
-    cmi_hwloc_topology_set_flags(topology, cmi_hwloc_topology_get_flags(topology) | HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM);
-    cmi_hwloc_topology_load(topology);
-
-    depth = cmi_hwloc_get_type_depth(topology, HWLOC_OBJ_PU);
-    CmiHwlocTopologyLocal.total_num_pus = depth != HWLOC_TYPE_DEPTH_UNKNOWN ? cmi_hwloc_get_nbobjs_by_depth(topology, depth) : 1;
-
-    cmi_hwloc_topology_destroy(topology);
+    depth = cmi_hwloc_get_type_depth(legacy_topology, HWLOC_OBJ_PU);
+    CmiHwlocTopologyLocal.total_num_pus =
+      depth != HWLOC_TYPE_DEPTH_UNKNOWN ? cmi_hwloc_get_nbobjs_by_depth(legacy_topology, depth) : 1;;
```
Comment (stwhite91) [inline src/conv-core/cpuaffinity.C]: Double semi-colon at end of line

File: src/conv-core/cpuaffinity.C [hunk 3]
```diff
@@ -1097,6 +1036,13 @@ void CmiInitCPUAffinityUtil(void){
     }
 */
 #endif
+
+  cmi_hwloc_topology_init(&topology);
+  cmi_hwloc_topology_load(topology);
+
+  cmi_hwloc_topology_init(&legacy_topology);
+  cmi_hwloc_topology_set_flags(legacy_topology, cmi_hwloc_topology_get_flags(legacy_topology) | HWLOC_TOPOLOGY_FLAG_INCLUDE_DISALLOWED);
```
Comment (matthiasdiener) [inline src/conv-core/cpuaffinity.C]: the flag was `HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM` before. is this intentional?

File: src/conv-core/cpuaffinity.C (hunk 1 above)
Comment (evan-charmworks) [reply src/conv-core/cpuaffinity.C]: Done

File: src/conv-core/cpuaffinity.C (hunk 2 above)
Comment (evan-charmworks) [reply src/conv-core/cpuaffinity.C]: Fixed

File: src/conv-core/cpuaffinity.C (hunk 3 above)
Comment (evan-charmworks) [reply src/conv-core/cpuaffinity.C]: Yes, the `HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM` was deprecated and renamed in hwloc 2.1.0.

**Comment by jcphill** [comment]
6.10.0 branch now fails with
cpuaffinity.C: In function 'void CmiInitHwlocTopology()':
cpuaffinity.C:72:99: error: 'HWLOC_TOPOLOGY_FLAG_INCLUDE_DISALLOWED' was not declared in this scope
     cmi_hwloc_topology_set_flags(legacy_topology, cmi_hwloc_topology_get_flags(legacy_topology) | HWLOC_TOPOLOGY_FLAG_INCLUDE_DISALLOWED);
                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Fatal Error by charmc in directory /tmp/oct2019/charm/netlrts-linux-x86_64/tmp
   Command g++ -DCMK_GFORTRAN -I../bin/../include -D__CHARMC__=1 -I. -O3 -U_FORTIFY_SOURCE -fno-stack-protector -fno-lifetime-dse -c cpuaffinity.C -o cpuaffinity.o returned error code 1

**Comment by evan-charmworks** [comment]
> 6.10.0 branch now fails

Pushed a fix.

**Pull request merged**